File: test/reference-output/generics.fs
module Generics

open Thoth.Json.Net

type UsingGenerics =
    {
        Field1: Basic.Maybe<string>
        Field2: Basic.Either<string, uint32>
    }

    static member Decoder: Decoder<UsingGenerics> =
        Decode.object (fun get ->
            {
                Field1 = get.Required.Field "field1" (Basic.Maybe.Decoder Decode.string)
                Field2 = get.Required.Field "field2" (Basic.Either.Decoder Decode.string Decode.uint32)
            }
        )

    static member Encoder value =
        Encode.object
            [
                "field1", (Basic.Maybe.Encoder Encode.string) value.Field1
                "field2", (Basic.Either.Encoder Encode.string Encode.uint32) value.Field2
            ]

type UsingOwnGenerics<'t> =
    {
        Field1: Basic.Maybe<'t>
    }

    static member Decoder decodeT: Decoder<UsingOwnGenerics<'t>> =
        Decode.object (fun get ->
            {
                Field1 = get.Required.Field "field1" (Basic.Maybe.Decoder decodeT)
            }
        )

    static member Encoder encodeT value =
        Encode.object
            [
                "field1", (Basic.Maybe.Encoder encodeT) value.Field1
            ]

type KnownForMovie =
    {
        Media_type: string
        Poster_path: option<string>
        Id: uint32
        Title: option<string>
        Vote_average: float32
        Release_date: option<string>
        Overview: string
    }

    static member Decoder: Decoder<KnownForMovie> =
        Decode.object (fun get ->
            {
                Media_type = get.Required.Field "media_type" (GotynoCoders.decodeLiteralString "movie")
                Poster_path = get.Optional.Field "poster_path" Decode.string
                Id = get.Required.Field "id" Decode.uint32
                Title = get.Optional.Field "title" Decode.string
                Vote_average = get.Required.Field "vote_average" Decode.float32
                Release_date = get.Optional.Field "release_date" Decode.string
                Overview = get.Required.Field "overview" Decode.string
            }
        )

    static member Encoder value =
        Encode.object
            [
                "media_type", Encode.string "movie"
                "poster_path", Encode.option Encode.string value.Poster_path
                "id", Encode.uint32 value.Id
                "title", Encode.option Encode.string value.Title
                "vote_average", Encode.float32 value.Vote_average
                "release_date", Encode.option Encode.string value.Release_date
                "overview", Encode.string value.Overview
            ]

type KnownForShow =
    {
        Media_type: string
        Poster_path: option<string>
        Id: uint32
        Vote_average: float32
        Overview: string
        First_air_date: option<string>
        Name: option<string>
    }

    static member Decoder: Decoder<KnownForShow> =
        Decode.object (fun get ->
            {
                Media_type = get.Required.Field "media_type" (GotynoCoders.decodeLiteralString "tv")
                Poster_path = get.Optional.Field "poster_path" Decode.string
                Id = get.Required.Field "id" Decode.uint32
                Vote_average = get.Required.Field "vote_average" Decode.float32
                Overview = get.Required.Field "overview" Decode.string
                First_air_date = get.Optional.Field "first_air_date" Decode.string
                Name = get.Optional.Field "name" Decode.string
            }
        )

    static member Encoder value =
        Encode.object
            [
                "media_type", Encode.string "tv"
                "poster_path", Encode.option Encode.string value.Poster_path
                "id", Encode.uint32 value.Id
                "vote_average", Encode.float32 value.Vote_average
                "overview", Encode.string value.Overview
                "first_air_date", Encode.option Encode.string value.First_air_date
                "name", Encode.option Encode.string value.Name
            ]

type KnownFor =
    | KnownForKnownForShow of KnownForShow
    | KnownForKnownForMovie of KnownForMovie
    | KnownForString of string
    | KnownForF32 of float32

    static member KnownForKnownForShowDecoder: Decoder<KnownFor> =
        Decode.map KnownForKnownForShow KnownForShow.Decoder

    static member KnownForKnownForMovieDecoder: Decoder<KnownFor> =
        Decode.map KnownForKnownForMovie KnownForMovie.Decoder

    static member KnownForStringDecoder: Decoder<KnownFor> =
        Decode.map KnownForString Decode.string

    static member KnownForF32Decoder: Decoder<KnownFor> =
        Decode.map KnownForF32 Decode.float32

    static member Decoder: Decoder<KnownFor> =
        Decode.oneOf
            [
                KnownFor.KnownForKnownForShowDecoder
                KnownFor.KnownForKnownForMovieDecoder
                KnownFor.KnownForStringDecoder
                KnownFor.KnownForF32Decoder
            ]

    static member Encoder =
        function
        | KnownForKnownForShow payload ->
            KnownForShow.Encoder payload

        | KnownForKnownForMovie payload ->
            KnownForMovie.Encoder payload

        | KnownForString payload ->
            Encode.string payload

        | KnownForF32 payload ->
            Encode.float32 payload

type KnownForMovieWithoutTypeTag =
    {
        Poster_path: option<string>
        Id: uint32
        Title: option<string>
        Vote_average: float32
        Release_date: option<string>
        Overview: string
    }

    static member Decoder: Decoder<KnownForMovieWithoutTypeTag> =
        Decode.object (fun get ->
            {
                Poster_path = get.Optional.Field "poster_path" Decode.string
                Id = get.Required.Field "id" Decode.uint32
                Title = get.Optional.Field "title" Decode.string
                Vote_average = get.Required.Field "vote_average" Decode.float32
                Release_date = get.Optional.Field "release_date" Decode.string
                Overview = get.Required.Field "overview" Decode.string
            }
        )

    static member Encoder value =
        Encode.object
            [
                "poster_path", Encode.option Encode.string value.Poster_path
                "id", Encode.uint32 value.Id
                "title", Encode.option Encode.string value.Title
                "vote_average", Encode.float32 value.Vote_average
                "release_date", Encode.option Encode.string value.Release_date
                "overview", Encode.string value.Overview
            ]

type KnownForShowWithoutTypeTag =
    {
        Poster_path: option<string>
        Id: uint32
        Vote_average: float32
        Overview: string
        First_air_date: option<string>
        Name: option<string>
    }

    static member Decoder: Decoder<KnownForShowWithoutTypeTag> =
        Decode.object (fun get ->
            {
                Poster_path = get.Optional.Field "poster_path" Decode.string
                Id = get.Required.Field "id" Decode.uint32
                Vote_average = get.Required.Field "vote_average" Decode.float32
                Overview = get.Required.Field "overview" Decode.string
                First_air_date = get.Optional.Field "first_air_date" Decode.string
                Name = get.Optional.Field "name" Decode.string
            }
        )

    static member Encoder value =
        Encode.object
            [
                "poster_path", Encode.option Encode.string value.Poster_path
                "id", Encode.uint32 value.Id
                "vote_average", Encode.float32 value.Vote_average
                "overview", Encode.string value.Overview
                "first_air_date", Encode.option Encode.string value.First_air_date
                "name", Encode.option Encode.string value.Name
            ]

type KnownForEmbedded =
    | MovieStartingWithLowercase of KnownForMovieWithoutTypeTag
    | TvStartingWithLowercase of KnownForShowWithoutTypeTag

    static member MovieStartingWithLowercaseDecoder: Decoder<KnownForEmbedded> =
        Decode.object (fun get ->
            MovieStartingWithLowercase {
                Poster_path = get.Optional.Field "poster_path" Decode.string
                Id = get.Required.Field "id" Decode.uint32
                Title = get.Optional.Field "title" Decode.string
                Vote_average = get.Required.Field "vote_average" Decode.float32
                Release_date = get.Optional.Field "release_date" Decode.string
                Overview = get.Required.Field "overview" Decode.string
            }
        )

    static member TvStartingWithLowercaseDecoder: Decoder<KnownForEmbedded> =
        Decode.object (fun get ->
            TvStartingWithLowercase {
                Poster_path = get.Optional.Field "poster_path" Decode.string
                Id = get.Required.Field "id" Decode.uint32
                Vote_average = get.Required.Field "vote_average" Decode.float32
                Overview = get.Required.Field "overview" Decode.string
                First_air_date = get.Optional.Field "first_air_date" Decode.string
                Name = get.Optional.Field "name" Decode.string
            }
        )

    static member Decoder: Decoder<KnownForEmbedded> =
        GotynoCoders.decodeWithTypeTag
            "media_type"
            [|
                "movieStartingWithLowercase", KnownForEmbedded.MovieStartingWithLowercaseDecoder
                "tvStartingWithLowercase", KnownForEmbedded.TvStartingWithLowercaseDecoder
            |]

    static member Encoder =
        function
        | MovieStartingWithLowercase payload ->
            Encode.object
                [
                    "media_type", Encode.string "movieStartingWithLowercase"
                    "poster_path", Encode.option Encode.string payload.Poster_path
                    "id", Encode.uint32 payload.Id
                    "title", Encode.option Encode.string payload.Title
                    "vote_average", Encode.float32 payload.Vote_average
                    "release_date", Encode.option Encode.string payload.Release_date
                    "overview", Encode.string payload.Overview
                ]

        | TvStartingWithLowercase payload ->
            Encode.object
                [
                    "media_type", Encode.string "tvStartingWithLowercase"
                    "poster_path", Encode.option Encode.string payload.Poster_path
                    "id", Encode.uint32 payload.Id
                    "vote_average", Encode.float32 payload.Vote_average
                    "overview", Encode.string payload.Overview
                    "first_air_date", Encode.option Encode.string payload.First_air_date
                    "name", Encode.option Encode.string payload.Name
                ]

type KnownForEmbeddedWithUpperCase =
    | Movie of KnownForMovieWithoutTypeTag
    | Tv of KnownForShowWithoutTypeTag

    static member MovieDecoder: Decoder<KnownForEmbeddedWithUpperCase> =
        Decode.object (fun get ->
            Movie {
                Poster_path = get.Optional.Field "poster_path" Decode.string
                Id = get.Required.Field "id" Decode.uint32
                Title = get.Optional.Field "title" Decode.string
                Vote_average = get.Required.Field "vote_average" Decode.float32
                Release_date = get.Optional.Field "release_date" Decode.string
                Overview = get.Required.Field "overview" Decode.string
            }
        )

    static member TvDecoder: Decoder<KnownForEmbeddedWithUpperCase> =
        Decode.object (fun get ->
            Tv {
                Poster_path = get.Optional.Field "poster_path" Decode.string
                Id = get.Required.Field "id" Decode.uint32
                Vote_average = get.Required.Field "vote_average" Decode.float32
                Overview = get.Required.Field "overview" Decode.string
                First_air_date = get.Optional.Field "first_air_date" Decode.string
                Name = get.Optional.Field "name" Decode.string
            }
        )

    static member Decoder: Decoder<KnownForEmbeddedWithUpperCase> =
        GotynoCoders.decodeWithTypeTag
            "media_type"
            [|
                "Movie", KnownForEmbeddedWithUpperCase.MovieDecoder
                "Tv", KnownForEmbeddedWithUpperCase.TvDecoder
            |]

    static member Encoder =
        function
        | Movie payload ->
            Encode.object
                [
                    "media_type", Encode.string "Movie"
                    "poster_path", Encode.option Encode.string payload.Poster_path
                    "id", Encode.uint32 payload.Id
                    "title", Encode.option Encode.string payload.Title
                    "vote_average", Encode.float32 payload.Vote_average
                    "release_date", Encode.option Encode.string payload.Release_date
                    "overview", Encode.string payload.Overview
                ]

        | Tv payload ->
            Encode.object
                [
                    "media_type", Encode.string "Tv"
                    "poster_path", Encode.option Encode.string payload.Poster_path
                    "id", Encode.uint32 payload.Id
                    "vote_average", Encode.float32 payload.Vote_average
                    "overview", Encode.string payload.Overview
                    "first_air_date", Encode.option Encode.string payload.First_air_date
                    "name", Encode.option Encode.string payload.Name
                ]

File: test/reference-output/generics.kt
package org.gotynoOutput

import com.fasterxml.jackson.annotation.*
import com.fasterxml.jackson.module.kotlin.*
import com.fasterxml.jackson.databind.annotation.*
import com.fasterxml.jackson.databind.*
import com.fasterxml.jackson.core.*
import com.fasterxml.jackson.databind.deser.std.*
import java.text.ParseException
import java.math.BigInteger
import kotlinx.serialization.Serializable
import org.gotynoDeclarations.BigIntegerSerializer

import org.gotynoOutput.Basic
import org.gotynoOutput.HasGeneric

class Generics {
@Serializable
data class UsingGenerics(
    @get:JsonProperty("field1")
    val field1: Basic.Maybe<String>,
    @get:JsonProperty("field2")
    val field2: Basic.Either<String, Int>
) : java.io.Serializable {
    companion object {
        fun create(field1: Basic.Maybe<String>, field2: Basic.Either<String, Int>): UsingGenerics {
            return UsingGenerics(field1 = field1, field2 = field2)
        }
    }
}

@Serializable
data class UsingOwnGenerics<T>(
    @get:JsonProperty("field1")
    val field1: Basic.Maybe<T>
) : java.io.Serializable {
    companion object {
        fun <T> create(field1: Basic.Maybe<T>): UsingOwnGenerics<T> {
            return UsingOwnGenerics(field1 = field1)
        }
    }
}

@Serializable
data class KnownForMovie(
    @get:JsonProperty("poster_path")
    val poster_path: String?,
    @get:JsonProperty("id")
    val id: Int,
    @get:JsonProperty("title")
    val title: String?,
    @get:JsonProperty("vote_average")
    val vote_average: Float,
    @get:JsonProperty("release_date")
    val release_date: String?,
    @get:JsonProperty("overview")
    val overview: String,
    @get:JsonProperty("media_type")
    val media_type: String = "movie"
) : java.io.Serializable {
    companion object {
        fun create(id: Int, vote_average: Float, overview: String, poster_path: String? = null, title: String? = null, release_date: String? = null, media_type: String = "movie"): KnownForMovie {
            return KnownForMovie(media_type = media_type, poster_path = poster_path, id = id, title = title, vote_average = vote_average, release_date = release_date, overview = overview)
        }
    }
}

@Serializable
data class KnownForShow(
    @get:JsonProperty("poster_path")
    val poster_path: String?,
    @get:JsonProperty("id")
    val id: Int,
    @get:JsonProperty("vote_average")
    val vote_average: Float,
    @get:JsonProperty("overview")
    val overview: String,
    @get:JsonProperty("first_air_date")
    val first_air_date: String?,
    @get:JsonProperty("name")
    val name: String?,
    @get:JsonProperty("media_type")
    val media_type: String = "tv"
) : java.io.Serializable {
    companion object {
        fun create(id: Int, vote_average: Float, overview: String, poster_path: String? = null, first_air_date: String? = null, name: String? = null, media_type: String = "tv"): KnownForShow {
            return KnownForShow(media_type = media_type, poster_path = poster_path, id = id, vote_average = vote_average, overview = overview, first_air_date = first_air_date, name = name)
        }
    }
}

@JsonDeserialize(using = KnownFor.Deserializer::class)
@Serializable
sealed class KnownFor : java.io.Serializable {
    @Serializable
    data class _KnownForShow(@JsonValue(true) val data: KnownForShow) : KnownFor()
    @Serializable
    data class _KnownForMovie(@JsonValue(true) val data: KnownForMovie) : KnownFor()
    @Serializable
    data class _String(@JsonValue(true) val data: String) : KnownFor()
    @Serializable
    data class _Float(@JsonValue(true) val data: Float) : KnownFor()

    class Deserializer : StdDeserializer<KnownFor>(KnownFor::class.java) {
        override fun deserialize(p: JsonParser, ctxt: DeserializationContext): KnownFor {
            val text = ctxt.readTree(p).toString()
            val mapper = jacksonObjectMapper()

            try { return _KnownForShow(mapper.readValue(text)) } catch (_: Exception)  {}
            try { return _KnownForMovie(mapper.readValue(text)) } catch (_: Exception)  {}
            try { return _Float(mapper.readValue(text)) } catch (_: Exception)  {}
            try { return _String(mapper.readValue(text)) } catch (_: Exception)  {}
            throw ParseException("Could not deserialize to class 'KnownFor'", 0)
        }
    }
}

@Serializable
data class KnownForMovieWithoutTypeTag(
    @get:JsonProperty("poster_path")
    val poster_path: String?,
    @get:JsonProperty("id")
    val id: Int,
    @get:JsonProperty("title")
    val title: String?,
    @get:JsonProperty("vote_average")
    val vote_average: Float,
    @get:JsonProperty("release_date")
    val release_date: String?,
    @get:JsonProperty("overview")
    val overview: String
) : java.io.Serializable {
    companion object {
        fun create(id: Int, vote_average: Float, overview: String, poster_path: String? = null, title: String? = null, release_date: String? = null): KnownForMovieWithoutTypeTag {
            return KnownForMovieWithoutTypeTag(poster_path = poster_path, id = id, title = title, vote_average = vote_average, release_date = release_date, overview = overview)
        }
    }
}

@Serializable
data class KnownForShowWithoutTypeTag(
    @get:JsonProperty("poster_path")
    val poster_path: String?,
    @get:JsonProperty("id")
    val id: Int,
    @get:JsonProperty("vote_average")
    val vote_average: Float,
    @get:JsonProperty("overview")
    val overview: String,
    @get:JsonProperty("first_air_date")
    val first_air_date: String?,
    @get:JsonProperty("name")
    val name: String?
) : java.io.Serializable {
    companion object {
        fun create(id: Int, vote_average: Float, overview: String, poster_path: String? = null, first_air_date: String? = null, name: String? = null): KnownForShowWithoutTypeTag {
            return KnownForShowWithoutTypeTag(poster_path = poster_path, id = id, vote_average = vote_average, overview = overview, first_air_date = first_air_date, name = name)
        }
    }
}

@Serializable
@JsonTypeInfo(
    use = JsonTypeInfo.Id.NAME,
    include = JsonTypeInfo.As.EXISTING_PROPERTY,
    property = "media_type"
)
sealed class KnownForEmbedded : java.io.Serializable {
    @Serializable
    @JsonTypeName("movieStartingWithLowercase")
    data class MovieStartingWithLowercase(@JsonValue(true) val data: KnownForMovieWithoutTypeTag) : KnownForEmbedded(), java.io.Serializable {
        val media_type = "movieStartingWithLowercase"
    }

    @Serializable
    @JsonTypeName("tvStartingWithLowercase")
    data class TvStartingWithLowercase(@JsonValue(true) val data: KnownForShowWithoutTypeTag) : KnownForEmbedded(), java.io.Serializable {
        val media_type = "tvStartingWithLowercase"
    }
}

@Serializable
@JsonTypeInfo(
    use = JsonTypeInfo.Id.NAME,
    include = JsonTypeInfo.As.EXISTING_PROPERTY,
    property = "media_type"
)
sealed class KnownForEmbeddedWithUpperCase : java.io.Serializable {
    @Serializable
    @JsonTypeName("Movie")
    data class Movie(@JsonValue(true) val data: KnownForMovieWithoutTypeTag) : KnownForEmbeddedWithUpperCase(), java.io.Serializable {
        val media_type = "Movie"
    }

    @Serializable
    @JsonTypeName("Tv")
    data class Tv(@JsonValue(true) val data: KnownForShowWithoutTypeTag) : KnownForEmbeddedWithUpperCase(), java.io.Serializable {
        val media_type = "Tv"
    }
}
}
